tutorial/write-custom-widget.md

```diff
@@ -107,3 +107,178 @@ Take care not to keep any important state in your renderer - animation tickers
+
+```go
+type MyWidget struct {
+	widget.BaseWidget           // Inherit from BaseWidget
+	minSize           fyne.Size // The minimum size of the widget
+	text              string	// The text to display in the widget
+}
+```
+
+To construct the widget create a New<widgetname> function. The example we are using is called MyWidget so by convention the following constructor function name would be used:
+
+```go
+func NewMyWidget(W, H float32, text string) *MyWidget {
+	w := &MyWidget{ 							 // Initialise the widget storage
+		text:    text,                           // Save the text
+		minSize: fyne.Size{Width: W, Height: H}, // Set the size
+	}
+	w.ExtendBaseWidget(w) // Required to initialise the base widget
+	return w
+}
+```
+
+We need to override the MinSize() function so we can return the minimum size of our widget. 
+
+```go
+func (w *MyWidget) MinSize() fyne.Size {
+	return w.minSize
+}
+func (w *MyWidget) Resize(s fyne.Size) { // See note below
+	w.BaseWidget.Resize(s)
+}
+```
+
+Note : It is not necessary to override the Resize function (shown above) but if you need to change the Resize behaviour of your widget and you add a ```Resize(s fyne.Size)``` function then you **must** include the ```w.BaseWidget.Resize(s)```. in the code.
+
+The final function to add is the ```CreateRenderer() fyne.WidgetRenderer``` function:
+
+```go
+func (w *MyWidget) CreateRenderer() fyne.WidgetRenderer {
+	return newMyWidgetRenderer(w.minSize, w.text)
+}
+```
+
+**Warning**: Do not keep a reference to the widget renderer in the widget (or any other part of your application). The fyne application caches the renderer and may destroy it at any time.
+
+The fyne application will call ```CreateRenderer()``` again if it requires a new renderer instance.
+
+It is ok for the widget renderer to refer to the widget but not the other way round.
+
+The widget is now almost complete except for the ```newMyWidgetRenderer()``` function which creates your new widget renderer. 
+
+#### Create a Widget Renderer
+
+The widget renderer is how we tell the fyne application what to draw and how to lay it out.
+
+The widget renderer has it's own state and functions (see ```fyne.WidgetRenderer``` interface) but this should only be concerned with drawing (rendering) the widget.
+
+The renderer is responsible for:
+
+1) The objects that are to be drawn. These are returned by the ```Objects() []fyne.CanvasObject``` function. Notice that you can return any object as long as it implements the ```fyne.CanvasObject``` interface.
+2) Each object returned by the Objects() function will need to be created by the widget renderer and initialised before returning.
+3) The fyne application will then call the renderers Layout() function. This is where the size and position of each object is detirmined.
+
+Our example renderer could be stored in the following struct:
+
+```
+type myWidgetRenderer struct {
+	minSize fyne.Size // The size of the widget
+	rect    *canvas.Rectangle
```

Comment:
Why not use a semantic name like "background" or "bg" so it is clearer what these variables are.
Repetition is rather anti-go.